src/Compilers/CSharp/Portable/Emitter/Model/PEModuleBuilder.cs

```diff
@@ -208,6 +208,104 @@ internal sealed override IEnumerable<Cci.INestedTypeDefinition> GetSynthesizedNe
+                                {
+                                    case SymbolKind.Namespace:
+                                    case SymbolKind.NamedType:
+                                        namespacesAndTypesToProcess.Push((NamespaceOrTypeSymbol)member);
+                                        break;
+                                    default:
+                                        throw ExceptionUtilities.UnexpectedValue(member.Kind);
+                                }
+                            }
+                        }
+                        break;
+                    case SymbolKind.NamedType:
+                        var typeDefinition = (Cci.ITypeDefinition)symbol.GetCciAdapter();
+                        GetDocumentsForMethods(methodDocumentList, typeDefinition, context);
+
+                        foreach (var loc in symbol.Locations)
+                        {
+                            if (!loc.IsInSource)
+                            {
+                                continue;
+                            }
+
+                            var span = loc.GetLineSpan();
+                            var debugDocument = DebugDocumentsBuilder.TryGetDebugDocument(span.Path, basePath: null);
+
+                            // If we have a debug document that is already referenced by method debug info then we don't need to include it
+                            if (debugDocument is not null && !methodDocumentList.Contains(debugDocument))
+                            {
+                                debugDocuments.Add(debugDocument);
+                            }
+                        }
+
+                        if (debugDocuments.Count > 0)
+                        {
+                            result.Add((typeDefinition, debugDocuments.ToImmutable()));
+                        }
+
+                        debugDocuments.Clear();
+                        methodDocumentList.Clear();
+                        break;
+                    default:
+                        throw ExceptionUtilities.UnexpectedValue(symbol.Kind);
+                }
+            }
+
+            namespacesAndTypesToProcess.Free();
+            debugDocuments.Free();
+            methodDocumentList.Free();
+            return result.ToImmutableAndFree();
+        }
+
+        /// <summary>
+        /// Gets a list of documents from the method definitions in this type, and also outputs whether there are
+        /// any methods that don't have document info (ie don't have any IL)
+        /// </summary>
+        private static void GetDocumentsForMethods(PooledHashSet<Cci.DebugSourceDocument> documentList, Cci.ITypeDefinition typeDefinition, EmitContext context)
+        {
+            var typeMethods = typeDefinition.GetMethods(context);
+
+            foreach (var method in typeMethods)
+            {
+                var body = method.GetBody(context);
+                if (body is null || body.SequencePoints.IsEmpty)
```

Comment:
>body.SequencePoints.IsEmpty [](http://example.com/codeflow?start=36&length=27)

This check seems redundant since we're looping over `body.SequencePoints` below.